drivers/mspi/mspi_nrfe.c

```diff
@@ -0,0 +1,644 @@
+ */
+static int api_transceive(const struct device *dev, const struct mspi_dev_id *dev_id,
+			  const struct mspi_xfer *req)
+{
+	(void)dev_id;
+	struct mspi_nrfe_data *drv_data = dev->data;
+	uint32_t packets_done = 0;
+	int rc;
+
+	/* TODO: add support for asynchronous transfers */
+	if (req->async) {
+		return -ENOTSUP;
+	}
+
+	if (req->num_packet == 0 || !req->packets ||
+	    req->timeout > CONFIG_MSPI_COMPLETION_TIMEOUT_TOLERANCE) {
+		return -EFAULT;
+	}
+
+	drv_data->xfer = *req;
+
+	rc = send_config(NRFE_MSPI_CONFIG_XFER, (void *)&drv_data->xfer, sizeof(struct mspi_xfer));
+	if (rc < 0) {
+		LOG_ERR("Send xfer config error: %d", rc);
+		return rc;
+	}
+
+	while (packets_done < drv_data->xfer.num_packet) {
+		rc = start_next_packet(&drv_data->xfer, packets_done);
+		if (rc < 0) {
+			LOG_ERR("Start next packet error: %d", rc);
+			return rc;
+		}
+		++packets_done;
+	}
+
+	return 0;
+}
+
+#if CONFIG_PM_DEVICE
+/**
+ * @brief Callback function to handle power management actions.
+ *
+ * This function is responsible for handling power management actions
+ * such as suspend and resume for the given device. It performs the
+ * necessary operations when the device is requested to transition
+ * between different power states.
+ *
+ * @param dev Pointer to the device structure.
+ * @param action The power management action to be performed.
+ *
+ * @retval 0 If successful.
+ * @retval -ENOTSUP If the action is not supported.
+ */
+static int dev_pm_action_cb(const struct device *dev, enum pm_device_action action)
+{
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		/* TODO: Handle PM suspend state */
+		break;
+	case PM_DEVICE_ACTION_RESUME:
+		/* TODO: Handle PM resume state */
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+#endif
+
+/**
+ * @brief Initialize the MSPI NRFE driver.
+ *
+ * This function initializes the MSPI NRFE driver. It is responsible for
+ * setting up the hardware and registering the IPC endpoint for the
+ * driver.
+ *
+ * @param dev Pointer to the device structure for the MSPI NRFE driver.
+ *
+ * @retval 0 If successful.
+ * @retval -errno If an error occurs.
+ */
+static int nrfe_mspi_init(const struct device *dev)
+{
+	int ret;
+	const struct device *ipc_instance = DEVICE_DT_GET(DT_NODELABEL(ipc0));
+	const struct mspi_nrfe_config *drv_cfg = dev->config;
+	const struct mspi_dt_spec spec = {
+		.bus = dev,
+		.config = drv_cfg->mspicfg,
+	};
+
+	ret = pinctrl_apply_state(drv_cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret) {
+		return ret;
+	}
+
+	ret = ipc_service_open_instance(ipc_instance);
+	if ((ret < 0) && (ret != -EALREADY)) {
+		LOG_ERR("ipc_service_open_instance() failure");
+		return ret;
+	}
+
+	ret = ipc_service_register_endpoint(ipc_instance, &ep, &ep_cfg);
+	if (ret < 0) {
+		LOG_ERR("ipc_service_register_endpoint() failure");
+		return ret;
+	}
+
+	/* Wait for ep to be bounded */
+	while (!atomic_test_and_clear_bit(&ipc_atomic_sem, NRFE_MSPI_EP_BOUNDED)) {
+	}
```

Comment:
Since this is app core side driver, I'd expect it to be mostly built with multithreading enabled.
So using semaphore here when multithreading is enabled would be better.